osquery/core/database/rocksdb_database.cpp

```diff
@@ -0,0 +1,388 @@
+  batch_write_options_ = rocksdb::WriteOptions();
+  batch_write_options_.disableWAL = true;
+  return openInternal(options, db_path);
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::checkDbConnection() {
+  if (db_ == nullptr) {
+    return createError(DatabaseError::DbIsNotOpen, "Database is closed");
+  }
+  return Success();
+}
+
+Expected<std::string, DatabaseError> RocksdbDatabase::getRawBytes(
+    const std::string& domain, const std::string& key) {
+  auto handle = getHandle(domain);
+  if (handle) {
+    std::shared_ptr<Handle> handle_ptr = handle.take();
+    return getRawBytesInternal(handle_ptr.get(), key);
+  }
+  return handle.takeError();
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::putRawBytes(
+    const std::string& domain,
+    const std::string& key,
+    const std::string& value) {
+  auto handle = getHandle(domain);
+  if (handle) {
+    std::shared_ptr<Handle> handle_ptr = handle.take();
+    return putRawBytesInternal(handle_ptr.get(), key, value);
+  }
+  return handle.takeError();
+}
+
+Expected<std::string, DatabaseError> RocksdbDatabase::getRawBytesInternal(
+    Handle* handle, const std::string& key) {
+  std::string value = "";
+  auto status = db_->Get(default_read_options_, handle, key, &value);
+  if (status.IsNotFound()) {
+    return createError(DatabaseError::KeyNotFound, "Value not found");
+  }
+  if (status.IsCorruption()) {
+    in_panic_ = true;
+    auto corruption_error =
+        createError(RocksdbError::DatabaseIsCorrupted, status.ToString());
+    auto error =
+        createError(DatabaseError::Panic, "", std::move(corruption_error));
+    panic(error);
+    return std::move(error);
+  }
+  if (!status.ok()) {
+    return createError(DatabaseError::FailToReadData, status.ToString());
+  }
+  return value;
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::putRawBytesInternal(
+    Handle* handle, const std::string& key, const std::string& value) {
+  auto status = db_->Put(default_write_options_, handle, key, value);
+  if (!status.ok()) {
+    createError(DatabaseError::FailToWriteData, status.ToString());
+  }
+  return Success();
+}
+
+Expected<RocksdbDatabase::HandleRef, DatabaseError> RocksdbDatabase::getHandle(
+    const std::string& domain) {
+  if (BOOST_UNLIKELY(in_panic_)) {
+    return createError(DatabaseError::Panic, "Database is in panic mode :(");
+  }
+  if (BOOST_UNLIKELY(db_ == nullptr)) {
+    return createError(DatabaseError::DbIsNotOpen, "Database is closed");
+  }
+  auto handle = handles_map_.find(domain);
+  if (BOOST_UNLIKELY(handle == handles_map_.end())) {
+    return createError(DatabaseError::DomainNotFound,
+                       "Unknown database domain");
+  }
+  return handle->second;
+}
+
+Expected<std::string, DatabaseError> RocksdbDatabase::getString(
+    const std::string& domain, const std::string& key) {
+  auto result = getRawBytes(domain, key);
+  if (result) {
+    std::string result_str = result.take();
+#ifdef DEBUG
```

Comment:
Instead of using macro definition here, which eventually can turn this code out to spaghetti code, would you like to use `debug_only::validate(...)` ? 🤓